src/module/system/actions/base.ts

```diff
@@ -0,0 +1,120 @@
+import {
+    Action,
+    ActionCost,
+    ActionUseOptions,
+    ActionVariant,
+    ActionVariants,
+    ActionVariantUseOptions,
+} from "@system/actions/types";
+import { getActionGlyph, sluggify } from "@util";
+
+interface BaseActionVariantData {
+    cost?: ActionCost;
+    description?: string;
+    name: string;
+    slug?: string;
+    traits?: string[];
+}
+
+interface BaseActionData<ActionVariantDataType extends BaseActionVariantData = BaseActionVariantData> {
+    description: string;
+    name: string;
+    traits?: string[];
+    variants: { 0: ActionVariantDataType } & ActionVariantDataType[];
+}
+
+abstract class BaseActionVariant implements ActionVariant {
+    readonly #action: BaseAction<BaseActionVariantData, BaseActionVariant>;
+    readonly #cost?: ActionCost;
+    readonly #description?: string;
+    readonly #name?: string;
+    readonly #slug?: string;
+    readonly #traits?: string[];
```

Comment:
I don't know that we need this level of paranoia. The `readonly` type is sufficient to keep a contributor from trying to reassign them, and if someone is being silly in a macro, then that's on them.

Reply:
But the getters don't serve any purpose.